sklego/meta/grouped_transformer.py

```diff
@@ -0,0 +1,153 @@
+
+        if y:
+            grouped_transformers = {
+                # Fit a clone of the transformer to each group
+                group: clone(self.transformer).fit(X_value[indices, :], y[indices])
+                for group, indices in group_indices.items()
+            }
+        else:
+            grouped_transformers = {
+                group: clone(self.transformer).fit(X_value[indices, :])
+                for group, indices in group_indices.items()
+            }
+
+        return grouped_transformers
+
+    def __check_transformer(self):
+        if not hasattr(self.transformer, "transform"):
+            raise ValueError(
+                "The supplied transformer should have a 'transform' method"
+            )
+
+    def fit(self, X, y=None):
+        """
+        Fit the transformers to the groups in X
+
+        :param X: Array-like with at least two columns, of which at least one corresponds to groups defined in init,
+                  and the remaining columns represent the values to transform.
+        :param y: (Optional) target variable
+        """
+        self.__check_transformer()
+
+        X_value = self.__check_value_columns(X)
+        X_group = self.__check_grouping_columns(X)
+
+        self.fallback_ = None
+
+        if self.use_global_model:
+            self.fallback_ = clone(self.transformer).fit(X_value)
+
+        self.transformers_ = self.__fit_grouped_transformer(X_group, X_value)
+
+        return self
+
+    def __transform_single_group(self, group, X):
+        """Transform a single group by getting its transformer from the fitted dict"""
+        # Keep track of the original index such that we can sort in __transform_groups
+        index = X.index
+        try:
+            group_transformer = self.transformers_[group]
+        except KeyError:
+            if self.fallback_:
+                group_transformer = self.fallback_
+            else:
+                raise ValueError(
+                    f"Found new group {group} during transform with use_global_model = False"
+                )
+
+        return pd.DataFrame(group_transformer.transform(X)).set_index(index)
+
+    def __transform_groups(self, X_group: pd.DataFrame, X_value: np.array):
+        """Transform all groups"""
+        # Reset indices such that they are the same in X_group (reset in __check_grouping_columns),
+        # this way we can track the order of the result
+        X_value = pd.DataFrame(X_value).reset_index(drop=True)
+
+        # Make the groups based on the groups dataframe, use the indices on the values array
+        group_indices = X_group.groupby(X_group.columns.tolist()).indices
+
+        return (
+            pd.concat(
+                [
+                    self.__transform_single_group(group, X_value.loc[indices, :])
+                    for group, indices in group_indices.items()
+                ],
+                axis=0,
+            )
+            .sort_index()
+            .values
+        )
```

Comment:
Is the output here always a pandas dataframe? Even when the input is an nd-array?

Reply:
Yes, in this specific line because it is the result of `pd.concat`. In general because I use the indices to keep track of the order of rows over groups (to make sure that input and output are in the same order, not sorted by group).

Reply:
That we use a pandas dataframe as an intermediate result is fine, but it may be safer to respect the input of the user. If a user passes a numpy array into this transfromer they might expect a numpy array to also come out of it. I'm also not 100% sure if all the scikit-learn tools are 100% pandas compatible.

Reply:
Sorry - misread the question. The intermediate result is always a pandas dataframe, the end result is always a numpy array (because of the `.values`). If you want, I can easily adapt it such that the output is a dataframe if the input is. I didn't do it now because most (all?) sklego transformers also return a numpy array.

Reply:
If the output remains numpy I think that is fine for now. It's what all the scikit-learn models do anyway. Cool. I'll give it one big final review now.

Reply:
Ah yeah. I only now see the `.values`.